pkg/ruler/ruler_test.go

```diff
@@ -216,3 +227,324 @@ func compareRuleGroupDescToStateDesc(t *testing.T, expected *rules.RuleGroupDesc
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, LEAVING ruler-1": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.LEAVING)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.ACTIVE)
+			},
+
+			// LEAVING still has its rules.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, LEAVING ruler-2": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.LEAVING)
+			},
+
+			// No extra rules from LEAVING.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, JOINING ruler-1": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.JOINING)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.ACTIVE)
+			},
+
+			// JOINING ruler has no rules yet.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{},
+		},
+
+		"default sharding, JOINING ruler-2": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.JOINING)
+			},
+
+			// JOINING ruler will get the rules.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"shuffle sharding, single ruler": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyShuffle,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{0}, ring.ACTIVE)
+			},
+
+			expectedRulesForFuler1: allRules,
+		},
+
+		"shuffle sharding, multiple rulers, shard size 1": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyShuffle,
+			shuffleShardSize: 1,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{userToken(user1, 0) + 1, userToken(user2, 0) + 1, userToken(user3, 0) + 1}, ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{0}, ring.ACTIVE)
```

Comment:
Given this ruler has only the token 0, is this test case a fair one? The ruler 1 owns all required tokens for `allRules` so I suspect that would pass even if the shard size is 2. Am I wrong?

Reply:
With shard size 2, it was in fact second ruler that got all the rules. I've now modified the test (added group tokens to ruler2), and added same test with shardSize=2 to show that.